keps/sig-testing/20190118-breaking-apart-the-kubernetes-test-tarball.md

```diff
@@ -0,0 +1,221 @@
+
+- `test/e2e/testing-manifests/`
+- `test/images/`
+- `test/kubemark/`
+- `hack/` ([partially](https://github.com/kubernetes/kubernetes/blob/193f659a1cd454b93cbe1e7b1f13b77c21783461/hack/lib/golang.sh#L193-L197))
+
+We have two options for these:
+
+1. Continue to distribute as a separate tarball, either `kubernetes-test.tar.gz`,
+   or possibly something like `kubernetes-test-portable.tar.gz`.
+- This makes the distinction very clear vs. the binary artifacts
+- There's already some precedent, such as the `kubernetes-manifest.tar.gz` tarball
+- It slightly complicates downloading of test dependencies
+2. Duplicate these sources into each binary-specific tarball.
+- Simplifies test dependency distribution
+- Portable sources are small (as of v1.13.2, approximately 2.7MB uncompressed
+  or about 186KB compressed) so duplication isn't a huge worry
+- Complicates extraction of tarballs with existing scripts, since they assume
+  everything is platform-specific
+
+It's not immediately obvious which of these is the better approach based on the
+tradeoffs.
+
+### Updating dependencies on `kubernetes-test.tar.gz`
+
+Currently the CI workflows and `kubetest` use the `cluster/get-kube.sh` and
+`cluster/get-kube-binaries.sh` scripts to download all artifacts, and
+conveniently `get-kube-binaries.sh` is versioned with the release artifacts in
+`kubernetes.tar.gz`, so simply making `get-kube-binaries.sh` aware of the new
+tarballs should be sufficient for most CI and developer needs.
+
+Because the test tarball includes binaries used both on the host running tests
+(such as the `e2e` binary), as well as binaries which may run the nodes
+(`e2e.node`), we would need to make sure to download binary test artifacts
+targeting the host platform, node platform, and possibly server platform.
+
+A quick search reveals a few other uses of `kubernetes-test.tar.gz`, mostly in
+`cluster/`. We can update these to use the platform-specific tarballs, possibly
+with a fallback to the mondo-tarball if worried about versioning.
+
+#### Dependencies outside the Kubernetes organization
+
+Searching GitHub for references to `kubernetes-test.tar.gz` largely returns
+forks of the main kubernetes repository (including some very old forks,
+identifiable by the script `e2e-from-release.sh`). Since these forks are not
+likely to depend on upstream release artifacts, we can ignore them.
+
+The Samsung SDS CNCT kraken-lib repository has a reference to `kubernetes-test.tar.gz`
+in its [conformance test script](https://github.com/samsung-cnct/kraken-lib/blob/aceab16c316bafcdb1f542dc67876dd2e5279f6b/build-scripts/conformance-tests#L16),
+but this repo is also marked deprecated and read-only, and there have been no
+changes since July 2018.
+
+In vmware/simple-k8s-test-env, the `sk8.sh` file uses
+[`kubernetes-test.tar.gz`](https://github.com/vmware/simple-k8s-test-env/blob/master/sk8.sh#L4481),
+and this repo seems actively maintained, so we should make sure this continues
+working.
```

Comment:
cc @akutz

Reply:
Hi @bentheelder,

Thank you for CCing me. I replied below without noticing the CC.

This isn’t that simple. This repo was recently refactored, and our entire test workflow uses the old code, including images that reference the current tarball.

Reply:
How hard will it be to update the tarball name to include an architecture?

Reply:
Hi @BenTheElder,

Not very at all. See the primary comment thread below for the reason it's more complicated at the moment than it should be.

Reply:
Hi @BenTheElder / @ixdy / @spiffxp,

I've been able to successfully roll a patch for another breaking change into the old codebase, so I'm confident I can handle this one as well. The only other caveat for me is that my logic currently relies on the file path structure of the tarball (since I ignore non-Linux amd64 bits when I decompress the HTTP response with `tar`). I'll need to ensure some testing for any changes to the layout of the tarball in addition to its name.

Reply:
@akutz right, this proposes that the internal structure would change a bit; rather than the test binaries being under `platforms/{OS}/{ARCH}/` they'd be under `test/bin`, but the tarballs would also have a different name (being platform-specific).

Handling the different paths and names across releases is a bit unfortunate; I added some thoughts about this into the Risks/Mitigations section. PTAL.